### ISSUE TITLE
An EnvoyFilter can cause istiod to not start. unmarshalerDecoder: unknown field "filters"

### ISSUE THREAD

**Issue by rvennam** [istio/istio#27839]
**Bug description**
An EnvoyFilter can cause 1.7 istiod to not start.

[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
A bad EnvoyFilter shouldn't cause istiod to not start. This can result in a user taking down the control plane.

**Steps to reproduce the bug**
1. Apply 
```
cat <<EOF | kubectl apply -f -
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: rvennam-filter
  namespace: default
spec:
  filters:
  - filterConfig:
      inlineCode: |
        function envoy_on_response(response_handle)
        end
    filterName: envoy.lua
    filterType: HTTP
    listenerMatch:
      listenerType: GATEWAY
EOF
```
2. Restart istiod pods
```
kubectl delete pod -n istio-system -l app=istiod
```
3. Look at the logs of the new istiod pods which fail to start. 
```
2020-10-09T00:59:13.014058Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
2020-10-09T00:59:13.114181Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
2020-10-09T00:59:13.214113Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
2020-10-09T00:59:13.251149Z	error	k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1alpha3.EnvoyFilter: v1alpha3.EnvoyFilterList.Items: []v1alpha3.EnvoyFilter: v1alpha3.EnvoyFilter.v1alpha3.EnvoyFilter.Spec: unmarshalerDecoder: unknown field "filters" in v1alpha3.EnvoyFilter, error found in #10 byte of ...|TEWAY"}}]}},{"apiVer|..., bigger context ...|TTP","listenerMatch":{"listenerType":"GATEWAY"}}]}},{"apiVersion":"networking.istio.io/v1alpha3","ki|...
2020-10-09T00:59:13.314115Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
2020-10-09T00:59:13.414153Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
2020-10-09T00:59:13.514094Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
2020-10-09T00:59:13.614136Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
2020-10-09T00:59:13.714109Z	info	kube	controller "networking.istio.io/v1alpha3/EnvoyFilter" is syncing...
```

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
client version: 1.7.0
control plane version: 1.7.3
data plane version: 1.7.3 (8 proxies), 1.7.1 (8 proxies), 1.7.0 (1 proxies)

**How was Istio installed?**
Operator

**Environment where bug was observed (cloud vendor, OS, etc)**
IBM Cloud

**Comment by howardjohn** [comment]
I have fixed this on master, we can backport it

**Comment by howardjohn** [comment]
https://github.com/istio/tools/commit/ca126563c43ac0b5989465edb9fcf7f64e03f93f

**Comment by rvennam** [comment]
Great! Do you think there are any other improvements that can be made to protect istiod against _any_ bad resource?

**Comment by kyessenov** [comment]
cc @yangminzhu with regards to authorization policy, an unknown field should not be disregarded, but instead imply denial.

**Comment by howardjohn** [comment]
https://github.com/istio/tools/pull/1247 for 1.7

**Comment by hniehus** [comment]
Hi,

I'm facing the exact same issue when updating Istio on AWS EKS with istioctl. 
Istio update path: 1.5.10->1.6.8->1.7.3

Is there any ETA for a permanent fix? Or any workaround for the meantime to get Istio back running?

Thanks,
Harald

**Comment by howardjohn** [comment]
Proper fix is in 1.7.4+, releasing soon. In the mean time, delete the config that has an unknown field. Might need to restart istiod after

**Comment by hniehus** [comment]
Thanks John.
Please forgive my ignorance, I'm still learning to walk with Istio - how do I find out which  config I need to scan for the unknown field?

**Comment by howardjohn** [comment]
I think in the log it says what the unknown field is. So I would do
`kubectl get envoyfilter --all-namespaces -oyaml | grep <field-name>` or
similar to pinpoint it.

**Comment by hniehus** [comment]
Hi John,
it took me a while to get my head around this, but finally I found the evil: actually, in my case, it was not an Istio issue but with a custom EnvoyFilter a colleague of mine wrote. Maybe he just "borrowed" from an older example, but actually the content of that EnvoyFilter was not following the newer specs - in the yaml I found keywords like "WorkloadLablels" that seemed to be valid for Istio <=1.2 but I couldn't find in the later references (like Istio 1.7).

So I think I'll need to rewrite that one yaml file to match the newer specs.

Thanks,
Harald

**Comment by LukasAuerbeck** [comment]
> Proper fix is in 1.7.4+, releasing soon. In the mean time, delete the config that has an unknown field. Might need to restart istiod after

We encountered the same problem while migrating our clusters from the old helm/direct istioctl install to istio-operators. I think the config should still be valid since it matches the latest EnvoyFilter spec. Since the filter is used to configure certain security header (csp, xss-protection, x-frame, ...) we would prefer to keep the EnvoyFilter. Does release 1.7.4 already have a planned release date?

**Comment by howardjohn** [comment]
Fixed by https://github.com/istio/tools/pull/1247
